### DISCUSSION TITLE
Some code not executed in production build in Next 14

### DISCUSSION THREAD

**Discussion by jira-zz** [vercel/next.js#70266] [Help] [answered]
### Link to the code that reproduces this issue

https://github.com/jira-zz/next-logging-test

### To Reproduce

1. Build the app in prod - `next build`
2. Run the app - `node .next/standalone/server.js`
3. Access the home page - `localhost:3000`

### Current vs. Expected behavior

> tady1
> tady2

should be printed to console

only 

> tady1 

is.

But it is not just console.log. For example, writing to a file is also skipped, logging with winston and other.

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP Thu Jan 11 04:09:03 UTC 2024
  Available memory (MB): 7759
  Available CPU cores: 16
Binaries:
  Node: 22.6.0
  npm: 10.8.2
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.2.11 // Latest available version is detected (14.2.11).
  eslint-config-next: 14.2.9
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.6.2
Next.js Config:
  output: standalone
```

### Which area(s) are affected? (Select all that apply)

Pages Router, Runtime

### Which stage(s) are affected? (Select all that apply)

next build (local), next start (local), Other (Deployed)

### Additional context

Happens with production build on my local machine, also when deployed in docker container.
The issue seems to be in production build.
I have an app that works in dev but fails in production. At first, I tried to debug it by throwing in some log statements, and then I found they are not executed.

Downgrading to next 13 fixes the issue.

The function in question is called from server component.

**Comment by samcx** [comment]
@jira-zz This actually happens regardless of using `output: standalone`.

When you build your project, we're collecting Page data and generating static pages. 

```
70150-logging-test [main] p build && p start

> next-logging-test@0.1.0 build .../Vercel Next.js/70150-logging-test
> next build

  ▲ Next.js 15.0.0-canary.159

   Creating an optimized production build ...
 ✓ Compiled successfully
 ✓ Linting and checking validity of types
   Collecting page data  .tady1
 ✓ Collecting page data
   Generating static pages (0/5)  [=   ]tady1
tady2
 ✓ Generating static pages (5/5)
 ✓ Collecting build traces
 ✓ Finalizing page optimization

Route (app)                              Size     First Load JS
┌ ○ /                                    138 B          97.4 kB
└ ○ /_not-found                          894 B          98.2 kB
+ First Load JS shared by all            97.3 kB
  ├ chunks/70d2848c-d777040b18d752c2.js  52.9 kB
  ├ chunks/8-30040b63e733d969.js         42.6 kB
  └ other shared chunks (total)          1.85 kB


○  (Static)  prerendered as static content


> next-logging-test@0.1.0 start .../Vercel Next.js/70150-logging-test
> next start

  ▲ Next.js 15.0.0-canary.159
  - Local:        http://localhost:3000

 ✓ Starting...
 ✓ Ready in 152ms
tady1
```

Since your Index page is rendered statically, the result of `getData()` is cached, and the inner console.log doesn't get ran again (it only ran once during the build).

In Development, we always render pages on-demand, so users don't have to wait for, e.g., revalidation periods.

**Comment by jira-zz** [comment]
Thanks for the reply. 
In my situation, the function fetches data from db. The result can obviously vary. How do I handle such situation?
Sorry if this is a dumb question.

**Reply by samcx** [reply, marked as answer]
@jira-zz You can force it to be dynamic → https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamic.

```
export const dynamic = 'force-dynamic'
```

Or, you can add a revalidate if you don't want it to request new data every time → https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#revalidate.

```
export const revalidate = 10
```
